docs/modifyrepo.1
.TH "modifyrepo" "1" "2007 Dec 3" "Luke Macken" ""

.SH "NAME"
modifyrepo \- Modify a repomd (xml-rpm-metadata) repository

.SH "SYNOPSIS"
\fBmodifyrepo\fP [options] <input metadata> <output repodata>
.PP

.SH "DESCRIPTION"
\fBmodifyrepo\fP is a program that allows you to insert arbitrary metadata into a repomd (xml-based rpm metadata) repository.

.SH "OPTIONS"
.IP "\fB\-\-mdtype <mdtype>\fP"
Specify datatype of the metadata. If not specified, datatype will be derived from the filename.

.IP "\fB\-\-remove\fP"
Remove specified file from repodata.

.IP "\fB\-\-compress\fP"
Compress the new repodata before adding it to the repo. This is used by default.

.IP "\fB\-\-no-compress\fP"
Do not compress the new repodata before adding it to the repo.

.IP "\fB\-\-compress-type <compress-type>\fP"
Specify which compression type to use: gz, xz (may not be available), bz2.

.IP "\fB\-s, \-\-checksum <sumtype>\fP"
Specify the checksum type to use.

.IP "\fB\-\-unique-md-filenames\fP"
Include the file's checksum in the metadata filename, helps with HTTP caching (default).

.IP "\fB\-\-simple-md-filenames\fP"
Do not include the file's checksum in the metadata filename.

.IP "\fB\-\-version\fP"
Output version.

.IP "\fB\-h \-\-help\fP"
Show help menu.

.SH "EXAMPLES"
.PP
$ \fBmodifyrepo\fP \-\-mdtype=newmd metadata.xml /repository/repodata
.br
Wrote: /repository/repodata/metadata.xml.gz
           type = newmd
       location = repodata/metadata.xml.gz
       checksum = 1d7ee93db2964e7f85e07ec19b3204591da1050c
      timestamp = 1196716296.0
  open-checksum = 824d936dc7dfff029379797b311af0cc66af4115
.br
Wrote: /repository/repodata/repomd.xml

.PP 
.SH "SEE ALSO"
.I createrepo (8)

.PP 
.SH "AUTHORS"
.nf 
Luke Macken <lmacken@redhat.com>
Seth Vidal <skvidal@fedoraproject.org>
.fi 

.PP 
.SH "BUGS"
Any bugs which are found should be emailed to the mailing list:
rpm-metadata@lists.baseurl.org
.fi


ChangeLog
2014-01-28  Zdenek Pavlas <zpavlas@redhat.com>
	* createrepo --update: ignore cached rpm when checksum_type
	  has changed.

2014-01-16  Zdenek Pavlas <zpavlas@redhat.com>

	* catch IOErrors when loading pkglist. BZ 1044997
	* Add missing int() on mtime
	* Fix file/dir completions for names containing spaces or tabs

2013-11-28  Zdenek Pavlas <zpavlas@redhat.com>

	* Removing the workers=1 default breaks API users, put it back.
	* Make sure the "packages" attribute is always correct.
	* Clean up tempdir on failures.
	* Close lock file before unlink()
	* mark as 0.10.1

2013-10-18  Zdenek Pavlas <zpavlas@redhat.org>

	* mark as 0.10

2011-01-26  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec, createrepo/__init__.py: mark as 0.9.9

2011-01-26  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo
	* 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo:
	Override timestamp check on repos. for mergerepo (like repodiff).
	Add createrepo --workers (non)completion.   Add modifyrepo option
	completion.

2011-01-21  James Antill <james@and.org>

	* createrepo/merge.py: Override timestamp check on repos. for
	mergerepo (like repodiff).

2011-01-03  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: make sure when we want to look for rpms we
	say .rpm not rpm b/c with the latter we catch .drpm files, too. :(

2010-11-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: Add createrepo --workers (non)completion.

2010-11-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: Add modifyrepo option completion.

2010-10-08  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec, createrepo/__init__.py: - add yum 3.2.29 requirement b/c of the small change I needed to
	repoMDObject.py -  set it to use /usr/share/createrepo/worker.py

2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: remove libxml2 import from __init__.py :)


2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: make createrepo use the repomd/repodata
	mechanism from yum for making a repomd.xml which simplifies the code
	dramatically since we don't have to mess with xml in here.

2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* modifyrepo.py: fix up the usage output for modifyrepo

2010-09-10  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, worker.py: - make sure we handle remote_url pkgs correctly until we get the
	worker hooked up to handle them - if there are no pkgs to handle,
	don't launch workers with nothing to do. - give better output from
	the workers and have them obey -v/-q - everyone loves callbacks!

2010-09-09  Seth Vidal <skvidal@fedoraproject.org>

	* Makefile, createrepo/__init__.py, createrepo/utils.py,
	createrepo/yumbased.py, genpkgmetadata.py, worker.py: create a
	worker script for createrepo so createrepo can fork off N processes
	to handle the md gathering from pkgs. This should speed up results
	on systems which have been cpubound on the createrepo process.  If
	you're io bound it won't help you at all, and MAY make it worse.
	many misc issues to iron out here - not the least of which is the
	callback output and gathering stdout/stderr from the workers

2010-08-20  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: handle broken locking on nfs target dirs
	better if database is true.  - sqlite dbs don't like being made on
	locations without locking available. - if we know we're going to be
	creating dbs then we should attempt to lock before doing   anything
	else and bail out nicely if we can't get an exclusive lock

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec: require yum 3.2.28 due to the imports in
	modifyrepo

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* docs/modifyrepo.1: document --mdtype option

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* modifyrepo.py: - add option parsing for --mdtype - use the yum repomd objects to
	read/write the repomd.xml - add mdtype option to RepoMetadata.add()
	method

2010-06-11  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo
	* 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo:
	Don't use /usr/bin/env ... it's evil   --database is now the default
	for mergerepo too, have --no-database in completions instead.

2010-06-11  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: - add option to createrepo config to collapse libc.so.6 requires -
	this will only work with yum 3.2.28 and beyond

2010-06-03  James Antill <james@and.org>

	* modifyrepo.py: Don't use /usr/bin/env ... it's evil

2010-06-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: --database is now the default for mergerepo too,
	have --no-database in completions instead.

2010-06-01  Seth Vidal <skvidal@fedoraproject.org>

	* mergerepo.py: whoops - no-database shouldn't default to true!

2010-06-01  Seth Vidal <skvidal@fedoraproject.org>

	* mergerepo.py: add --no-database to mergrepo, too

2010-05-31  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: --database is now the default, have --no-database
	in completions instead.

2010-05-28  Seth Vidal <skvidal@fedoraproject.org>

	* docs/createrepo.8: update the docs for --no-database

2010-05-28  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, genpkgmetadata.py: make -d/--database  the
	default  add --no-database in case someone somewhere needs to do
	that

2010-04-26  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: fixme comments about try/excepting the
	database creation calls due to a weird issue with locks not working
	on a nfs mount and createreepo tracing back with a TypeError of all
	things

2010-04-21  Seth Vidal <skvidal@fedoraproject.org>


modifyrepo.py
#!/usr/bin/python
# This tool is used to manipulate arbitrary metadata in a RPM repository.
# Example:
#           ./modifyrepo.py updateinfo.xml myrepo/repodata
#           or
#           ./modifyrepo.py --remove updateinfo.xml myrepo/repodata
# or in Python:
#           >>> from modifyrepo import RepoMetadata
#           >>> repomd = RepoMetadata('myrepo/repodata')
#           >>> repomd.add('updateinfo.xml')
#           or
#           >>> repomd.remove('updateinfo.xml')
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# (C) Copyright 2006  Red Hat, Inc.
# Luke Macken <lmacken@redhat.com>
# modified by Seth Vidal 2008
# modified by Daniel Mach 2011

import os
import re
import sys
from createrepo import __version__
from createrepo.utils import checksum_and_rename, compressOpen, MDError
from createrepo.utils import _available_compression
from yum.misc import checksum, _available_checksums, AutoFileChecksums

from yum.repoMDObject import RepoMD, RepoMDError, RepoData
from xml.dom import minidom
from optparse import OptionParser
from cStringIO import StringIO


class RepoMetadata:

    def __init__(self, repo):
        """ Parses the repomd.xml file existing in the given repo directory. """
        self.repodir = os.path.abspath(repo)
        self.repomdxml = os.path.join(self.repodir, 'repomd.xml')
        self.compress_type = _available_compression[-1] # best available
        self.compress = True
        self.checksum_type = 'sha256'
        self.unique_md_filenames = True

        if not os.path.exists(self.repomdxml):
            raise MDError, '%s not found' % self.repomdxml

        try:
            self.repoobj = RepoMD(self.repodir)
            self.repoobj.parse(self.repomdxml)
        except RepoMDError, e:
            raise MDError, 'Could not parse %s' % self.repomdxml

    def _get_mdtype(self, mdname, mdtype=None):
        """ Get mdtype from existing mdtype or from a mdname. """
        if mdtype:
            return mdtype
        mdname = os.path.basename(mdname)
        if re.match(r'[0-9a-f]{32,}-', mdname):
            mdname = mdname.split('-', 1)[1]
        return mdname.split('.')[0]

    def _print_repodata(self, repodata):
        """ Print repodata details. """
        print "           type =", repodata.type
        print "       location =", repodata.location[1]
        print "       checksum =", repodata.checksum[1]
        print "      timestamp =", repodata.timestamp
        print "  open-checksum =", repodata.openchecksum[1]
        print "           size =", repodata.size
        print "      open-size =", repodata.opensize

    def _write_repomd(self):
        """ Write the updated repomd.xml. """
        outmd = file(self.repomdxml, 'w')
        outmd.write(self.repoobj.dump_xml())
        outmd.close()
        print "Wrote:", self.repomdxml

    def _remove_repodata_file(self, repodata):
        """ Remove a file specified in repodata location """
        try:
            fname = os.path.basename(repodata.location[1])
            os.remove(os.path.join(self.repodir, fname))
        except OSError, ex:
            if ex.errno != 2:
                # continue on a missing file
                raise MDError("could not remove file %s" % repodata.location[1])

    def add(self, metadata, mdtype=None):
        """ Insert arbitrary metadata into this repository.
            metadata can be either an xml.dom.minidom.Document object, or
            a filename.
        """
        md = None
        if not metadata:
            raise MDError, 'metadata cannot be None'
        if isinstance(metadata, minidom.Document):
            md = metadata.toxml()
            mdname = 'updateinfo.xml'
            oldmd = AutoFileChecksums(StringIO(md), [self.checksum_type])
            oldmd.read()
        elif isinstance(metadata, str):
            if os.path.exists(metadata):
                mdname = os.path.basename(metadata)
                if mdname.split('.')[-1] in ('gz', 'bz2', 'xz'):
                    mdname = mdname.rsplit('.', 1)[0]
                    oldmd = compressOpen(metadata, mode='rb')
                else:
                    oldmd = file(metadata, 'r')
                oldmd = AutoFileChecksums(oldmd, [self.checksum_type])
                md = oldmd.read()
                oldmd.close()
            else:
                raise MDError, '%s not found' % metadata
        else:
            raise MDError, 'invalid metadata type'

        if not md and self.compress_type == 'xz':
            raise MDError, 'LZMA does not support compressing empty files'

        ## Compress the metadata and move it into the repodata
        mdtype = self._get_mdtype(mdname, mdtype)
        destmd = os.path.join(self.repodir, mdname)
        if self.compress:
            destmd += '.' + self.compress_type
            newmd = compressOpen(destmd, mode='wb', compress_type=self.compress_type)
        else:
            newmd = open(destmd, 'wb')
            
        newmd.write(md)
        newmd.close()
        print "Wrote:", destmd

        if self.unique_md_filenames:
            csum, destmd = checksum_and_rename(destmd, self.checksum_type)
        else:
            csum = checksum(self.checksum_type, destmd)
        base_destmd = os.path.basename(destmd)

        # Remove any stale metadata
        old_rd = self.repoobj.repoData.pop(mdtype, None)

        new_rd = RepoData()
        new_rd.type = mdtype
        new_rd.location = (None, 'repodata/' + base_destmd)
        new_rd.checksum = (self.checksum_type, csum)
        new_rd.size = str(os.stat(destmd).st_size)
        if self.compress:
            new_rd.openchecksum = (self.checksum_type, oldmd.checksums.hexdigests().popitem()[1])
            new_rd.opensize = str(oldmd.checksums.length)
        new_rd.timestamp = str(int(os.stat(destmd).st_mtime))
        self.repoobj.repoData[new_rd.type] = new_rd
        self._print_repodata(new_rd)
        self._write_repomd()

        if old_rd is not None and old_rd.location[1] != new_rd.location[1]:
            # remove the old file when overwriting metadata
            # with the same mdtype but different location
            self._remove_repodata_file(old_rd)

    def remove(self, metadata, mdtype=None):
        """ Remove metadata from this repository. """
        mdname = metadata
        mdtype = self._get_mdtype(mdname, mdtype)

        old_rd = self.repoobj.repoData.pop(mdtype, None)
        if old_rd is None:
            print "Metadata not found: %s" % mdtype
            return

        self._remove_repodata_file(old_rd)
        print "Removed:"
        self._print_repodata(old_rd)
        self._write_repomd()


def main(args):
    parser = OptionParser(version='modifyrepo version %s' % __version__)
    # query options
    parser.add_option("--mdtype", dest='mdtype',
                      help="specific datatype of the metadata, will be derived from the filename if not specified")
    parser.add_option("--remove", action="store_true",
                      help="remove specified file from repodata")
    parser.add_option("--compress", action="store_true", default=True,
                      help="compress the new repodata before adding it to the repo (default)")
    parser.add_option("--no-compress", action="store_false", dest="compress",
                      help="do not compress the new repodata before adding it to the repo")
    parser.add_option("--compress-type", dest='compress_type',
                      help="compression format to use")
    parser.add_option("-s", "--checksum", dest='sumtype',
        help="specify the checksum type to use")
    parser.add_option("--unique-md-filenames", dest="unique_md_filenames",
        help="include the file's checksum in the filename, helps with proxies",
        action="store_true")
    parser.add_option("--simple-md-filenames", dest="unique_md_filenames",
        help="do not include the file's checksum in the filename",
        action="store_false")
    parser.usage = "modifyrepo [options] [--remove] <input_metadata> <output repodata>"
    
    (opts, argsleft) = parser.parse_args(args)
    if len(argsleft) != 2:
        parser.print_usage()
        return 0
    metadata = argsleft[0]
    repodir = argsleft[1]
    try:
        repomd = RepoMetadata(repodir)
    except MDError, e:
        print "Could not access repository: %s" % str(e)
        return 1

    try:
        # try to extract defaults from primary entry
        md = repomd.repoobj.getData('primary')
        sumtype = md.checksum[0]
        name = os.path.basename(md.location[1])
        unique_md_filenames = re.match(r'[0-9a-f]{32,}-', name) != None
        compress_type = name.rsplit('.', 1)[1]
    except RepoMDError:
        sumtype = 'sha256'
        unique_md_filenames = True
        compress_type = 'gz'

    # apply defaults
    if opts.sumtype is None:
        opts.sumtype = sumtype
    if opts.unique_md_filenames is None:
        opts.unique_md_filenames = unique_md_filenames
    if opts.compress_type is None:
        opts.compress_type = compress_type

    repomd.checksum_type = opts.sumtype
    repomd.unique_md_filenames = opts.unique_md_filenames
    repomd.compress = opts.compress
    if opts.compress_type not in _available_compression:
        print "Compression %s not available: Please choose from: %s" % (opts.compress_type, ', '.join(_available_compression))
        return 1
    if opts.sumtype != 'sha' and opts.sumtype not in _available_checksums:
        print "Checksum %s not available: Please choose from: %s" % (opts.sumtype, ', '.join(_available_checksums))
        return 1
    repomd.compress_type = opts.compress_type

    # remove
    if opts.remove:
        try:
            repomd.remove(metadata, mdtype=opts.mdtype)
        except MDError, ex:
            print "Could not remove metadata: %s" % (metadata, str(ex))
            return 1
        return

    # add
    try:
        repomd.add(metadata, mdtype=opts.mdtype)
    except MDError, e:
        print "Could not add metadata from file %s: %s" % (metadata, str(e))
        return 1
    

if __name__ == '__main__':
    ret = main(sys.argv[1:])
    sys.exit(ret)
